runtime/vm/options.c

```diff
@@ -0,0 +1,198 @@
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include <stdio.h>
+#include <assert.h>
+#include "j9protos.h"
+#include "jvminitcommon.h"
+#include "options.h"
+
+
+/*
+ * Example -XX:+PrintFlagsFinal output: (Hotspot calls them "Flags" but OpenJ9 terminology is "options")
+ *
+ * [Global flags]
+ *    size_t MaxHeapSize                              = 4294967296                                {product} {ergonomic}
+ *  uint64_t MaxDirectMemorySize                      = 3758096384                                {product} {default}
+ *
+ * First column is the type of the option
+ * Second column is the name of the option
+ * Third column is the value of the option
+ * Fourth column is a category for the option: product, experimental, etc.
+ * Fifth column describes how the option's value was set: default setting, by command line, or via ergonomics (heuristics)
+ */
+
+
+/* Option types (first ten entries for compatibility) */
+#define TYPE_BOOL 0
+#define TYPE_CCSTR 1
+#define TYPE_CCSTRLIST 2
+#define TYPE_DOUBLE 3
+#define TYPE_INT 4
+#define TYPE_INTX 5
+#define TYPE_SIZET 6
+#define TYPE_UINT 7
+#define TYPE_UINT64T 8
+#define TYPE_UINTX 9
+
+static const char *option_type_string[] = {
+	"bool",
+	"ccstr",
+	"ccstrlist",
+	"double",
+	"int",
+	"intx",
+	"size_t",
+	"uint",
+	"uint64_t",
+	"uintx"
+};
+#define NUM_TYPE_STRINGS	(sizeof(option_type_string) / sizeof(option_type_string[0]))
+
+typedef union J9OptionValue {
+	UDATA v_uintx;
+	U_64 v_uint64;
+	size_t v_sizet;
+	double v_double;
+	char *v_ccstr;
+	char **v_ccstrlist;
+	U_32 v_uint;
+	IDATA v_intx;
+	I_8 v_bool;
+} J9OptionValue;
+
+#define CAT_PRODUCT		0
+#define CAT_ARCHPRODUCT		1
+#define CAT_PDPRODUCT		2
+#define CAT_PRODUCTRW		3
+#define CAT_DIAGNOSTIC		4
+#define CAT_EXPERIMENTAL	5
+#define CAT_MANAGEABLE		6
+
+static const char *option_category_string[] = {
+	"{product}",
+	"{ARCH product}",
+	"{pd product}",
+	"{product rw}",
+	"{diagnostic}",
+	"{experimental}",
+	"{manageable}",
+};
+#define NUM_CATEGORY_STRINGS	(sizeof(option_category_string) / sizeof(option_category_string[0]))
+
+#define SET_DEFAULT	0
+#define SET_COMMANDLINE	1
+#define SET_ERGONOMIC	2
+
+static const char *option_howset_string[] = {
```

Comment:
I was trying to faithfully implement what I could learn of PrintFlagsFinal from google links, since it's not well specified anywhere (at least, I couldn't find such a specification and I'm not going to look at Hotspot code). I think this is a case where we can add stuff as needed.